### ISSUE TITLE
webkitGenerateKeyRequest polyfill exception ( at least from WebOS 3.0 )

### ISSUE THREAD

**Issue by albertdaurell** [google/shaka-player#2915]
**Have you read the [FAQ](https://bit.ly/ShakaFAQ) and checked for duplicate open issues?**

Yes, no similar issues found

**What version of Shaka Player are you using?**

3.0.5

**Can you reproduce the issue with our latest release version?**

Yes, branch v3.0.x also contanis the issue

**Can you reproduce the issue with the latest code from `master`?**

Yes

**Are you using the demo app or your own custom app?**

Own custom app, but demo app will throw same exception ( I'll add an example )

**If custom app, can you reproduce the issue using our demo app?**

Yes, exception is thrown

**What browser and OS are you using?**

WebOS 3.0 ( 2016 )

**For embedded devices (smart TVs, etc.), what model and firmware version are you using?**

WebOS 3.0 ( 2016 )
firmware: 05.50.95
model: 43UH650V-ZB

**What are the manifest and license server URIs?**

Any DASH content + playready can be used ( See example I'll add )

**What did you do?**

Try to play any DASH playready DRM protected content

**What did you expect to happen?**

The content should play

**What actually happened?**

An exception is thrown from https://github.com/google/shaka-player/blob/acc4c0e2574874d2aaa9f2e7d177024d729d330b/lib/media/drm_engine.js#L1090-L1094

## ISSUE

Trying `Shaka Player` migration from `2.5.X` to `3.0.X` we faced this issue when 
`lib/polyfill/patchedmediakeys_webkit.js` is applied. A basic example:

```javascript
// Code in javascript
// Example from https://shaka-player-demo.appspot.com/demo/
var url = 'https://amssamples.streaming.mediaservices.windows.net/622b189f-ec39-43f2-93a2-201ac4e31ce1/BigBuckBunny.ism/manifest(format=mpd-time-csf)';
var license = 'https://amssamples.keydelivery.mediaservices.windows.net/PlayReady/';
// Intsall polyfills
shaka.polyfill.installAll();
// Create video element 
var video = document.querySelector('video');
if (!video) { 
  video = document.createElement('video'); 
  video.style.position = 'absolute';
  video.style.left = '0px';
  video.style.top = '0px';
  video.style.width = '100%';
  video.style.height = '100%';
  video.preload = 'auto';
  video.autoplay = true;
  document.body.appendChild(video);
}
// Override webkitGenerateKeyRequest in order to show more logs
if (!HTMLMediaElement.prototype._webkitGenerateKeyRequest) {
 HTMLMediaElement.prototype._webkitGenerateKeyRequest = HTMLMediaElement.prototype.webkitGenerateKeyRequest;
 HTMLMediaElement.prototype.webkitGenerateKeyRequest = function() {
   var args = [].slice.call(arguments);
   try {
     console.log('webkitGenerateKeyRequest call');
     this._webkitGenerateKeyRequest.apply(this, args);
   } catch(error) {
     console.log('webkitGenerateKeyRequest exception', error);
     throw error;
   }
 }
}
// Also show shaka logs
shaka.log.setLevel(shaka.log.Level.DEBUG);
// Create the player and play
var player = new shaka.Player(video); 
player.configure({ drm: { servers: { 'com.microsoft.playready': license } } });
player.attach(video);
player.load(url, 0)
  .catch(function(e) { console.log('Exception!', JSON.stringify(e)); });
```
And from this load an exception is thrown:

![image](https://user-images.githubusercontent.com/7487579/96251535-eb396500-0fb0-11eb-87ee-1eddbd99199b.png)

### Trace

Here is a "traced" explanation:

When load is called [onInitializeDrm_](https://github.com/google/shaka-player/blob/98df64c1af61490bdc926cfbf89a0d621e3a77cf/lib/player.js#L1656) ( from shaka.Player ) is called from `drm-engine`. And you replaced inside this method the call from:

&nbsp;&nbsp;`shaka 2.5` => [this.filterAllPeriods_(this.manifest_.periods);](https://github.com/google/shaka-player/blob/v2.5.x/lib/player.js#L1632)
&nbsp;&nbsp;vs 
&nbsp;&nbsp;`shaka 3.0` => [this.filterManifest_(this.manifest_);](https://github.com/google/shaka-player/blob/98df64c1af61490bdc926cfbf89a0d621e3a77cf/lib/player.js#L1707)

[filterManifest_](https://github.com/google/shaka-player/blob/98df64c1af61490bdc926cfbf89a0d621e3a77cf/lib/player.js#L4108) method is calling

[this.drmEngine_.newInitData(initData.initDataType, initData.initData);](https://github.com/google/shaka-player/blob/98df64c1af61490bdc926cfbf89a0d621e3a77cf/lib/player.js#L4148)

And newInitData from drmEngine method is calling:

[this.createTemporarySession_(initDataType, initData);](https://github.com/google/shaka-player/blob/98df64c1af61490bdc926cfbf89a0d621e3a77cf/lib/media/drm_engine.js#L552)

and this createTemporarySession_ method calls:

1. At first (this.mediaKeys_.createSession(...);](https://github.com/google/shaka-player/blob/98df64c1af61490bdc926cfbf89a0d621e3a77cf/lib/media/drm_engine.js#L1018-L1024)

which is creating this.media_ with an empty &lt;video src='about:blank' /&gt; see [createSession](https://github.com/google/shaka-player/blob/606457488f77e9cf5cac2c82e94bb4ce35d50e2a/lib/polyfill/patchedmediakeys_webkit.js#L396-L400)

2. And after that [session.generateRequest(initDataType, initData)](https://github.com/google/shaka-player/blob/98df64c1af61490bdc926cfbf89a0d621e3a77cf/lib/media/drm_engine.js#L1069)

which is calling finally to the video object method `webkitGenerateKeyRequest` (from the [generateRequest polyfill call](https://github.com/google/shaka-player/blob/606457488f77e9cf5cac2c82e94bb4ce35d50e2a/lib/polyfill/patchedmediakeys_webkit.js#L712)
and this `webkitGenerateKeyRequest` method is throwing a `DOMException` because correct video object is not being used 

```json
DOMException
{
  message: 'Failed to execute \'webkitGenerateKeyRequest\' on \'HTMLMediaElement\': No media has been loaded',
  name: 'InvalidStateError',
  code: 11,
  ...
}
```

And this exception is thrown as `CRITICAL` ( as it should )

We think the problem is because "attach" was not called, but since attach is called from load maybe there should be
applied a workaround in order to ensure this.media_ is correctly initialized when creating the session since the player
has been initialized from `new shaka.Player(video)` or `player.attach(video)` before calling `player.load`

Thank you very much

**Comment by joeyparrish** [comment]
On other platforms, there is no need to load media before generating a license request.  It is intentional that these two things are done in parallel.

We create a MediaSource object, construct a URL for it, and set that on the video element before we create a license request.  However, that may not take effect immediately, and we don't wait for the corresponding completion event before we start initializing DRM.

We don't have a WebOS device and therefore can't _directly_ support that platform.  However, here is a change you can try for us in `lib/player.js`, inside `onInitializeDrm_()`:

```js
    this.drmEngine_.configure(this.config_.drm);

    // ADD THESE LINES HERE
    // vvvvv
    if (shaka.util.Platform.isWebOS()) {
      await this.mediaSourceEngine_.open();
    }
    // ^^^^^

    await this.drmEngine_.initForPlayback(
        this.manifest_.variants,
        this.manifest_.offlineSessionIds);
```

If that resolves the issue, we can incorporate that workaround into our next release.

**Comment by albertdaurell** [comment]
@shaka-bot reopen

Hi, sorry for the late reply, we didn't had access to the WebOS device until today. 

**The fix did not worked**, we added logs in order to see the "workaround" was applied and .open is called but we were still not able to play.

Thx

**Comment by joeyparrish** [comment]
Unfortunately, we aren't able to debug this directly without a WebOS device.  We'll leave it open until someone can help diagnose and work around the issue.

**Comment by albertdaurell** [comment]
Hi @joeyparrish, since you can't debug I've been investigating and "trying to quickly understand the source code" :sweat_smile: and finally found a way to play, I modified onInitializeDrm_ BUT not with `await this.mediaSourceEngine_.open();` what I did is adding:

```javascript
    this.drmEngine_.configure(this.config_.drm);

    await this.drmEngine_.initForPlayback(
        this.manifest_.variants,
        this.manifest_.offlineSessionIds);
   
    // ADD THESE LINES HERE
    // vvvvv
    if (shaka.util.Platform.isWebOS()) {
      await this.drmEngine_.attach(has.mediaElement);
    }
    // ^^^^^

    // Now that we have drm information, filter the manifest (again) so that we
    // can ensure we only use variants with the selected key system.
    this.filterManifest_(this.manifest_);
```

I call `attach` after `initForPlayback` because order should be `initForPlayback` + `attach` like you're doing there:
https://github.com/google/shaka-player/blob/562a2d567b4c11a2fdd94e34b1ba3df751d0aeda/lib/player.js#L2016-L2018 Otherwise attach will not have `this.mediaKeys_` set and it will not store the `this.video_` object to be used from createSession

With this change player is working from a `WebOS 3.0`, also I tested this change from `WebOS 5.0` and `WebOS 4.5`

Thanks in advance

**Comment by joeyparrish** [comment]
Very interesting.  Good find!

It seems like it would be very reasonable to move `drmEngine_.attach()` from `onLoad_` to `onInitializeDrm_`.  Would you be interested to send a PR for this?

**Comment by avelad** [comment]
I created https://github.com/google/shaka-player/pull/3109 to resolve this

**Comment by albertdaurell** [comment]
Thx @avelad ( you were faster than me :sweat_smile: doing the PR ) we tested https://github.com/google/shaka-player/pull/3109 and issue get's fixed from `WebOS 3.0`, also I tested from `WebOS 4.5` + `WebOS 5.0` + `Tizen 2020` and player is working fine.

Thank you very much
